verl/workers/actor/dp_actor.py

```diff
@@ -41,17 +43,47 @@ def __init__(
         super().__init__(config)
         self.actor_module = actor_module
         self.actor_optimizer = actor_optimizer
+        self.use_remove_padding = self.config.get('use_remove_padding', False)
+        print(f'Actor use_remove_padding={self.use_remove_padding}')
 
-    def _forward_micro_batch(self, micro_batch, temperature):
+    def _forward_micro_batch(self, micro_batch, temperature) -> Tuple[torch.Tensor, torch.Tensor]:
         response_length = micro_batch['responses'].size(-1)
         with torch.autocast(device_type='cuda', dtype=torch.bfloat16):
-            output = self.actor_module(input_ids=micro_batch['input_ids'],
-                                       attention_mask=micro_batch['attention_mask'],
-                                       position_ids=micro_batch['position_ids'],
-                                       use_cache=False)  # prevent model thinks we are generating
-            logits = output.logits / temperature
-            logits = logits[:, -response_length - 1:-1]
-            log_probs = logprobs_from_logits(logits, micro_batch['responses'])
+            input_ids = micro_batch['input_ids']
+            batch_size, seqlen = input_ids.shape
+            attention_mask = micro_batch['attention_mask']
+            position_ids = micro_batch['position_ids']
+
+            if self.use_remove_padding:
```

Comment:
@PeterSH6 @vermouth1992 I am thinking if it's cleaner if the remove padding logit is also done in the monkey patch way, where the actor_module forward is patched and users always pass in the original input ids? 

Something like 
```
llama_forward_fn = LlamaModelForCausalLM.forward

def forward_remove_padding(...):
    input_rmpad = fn(input_ids)
    llama_forward_fn(xx_rmpad)

LlamaModelForCausalLM.forward = forward_remove_padding
```

Reply:
In my opinion, currently users can only need to pass the original input_ids to the data-proto. 
And we just need to unpad it before the model fwd.

If we use monkey-patch, we may need to maintain patches for all models.

Reply:
I agree that we should explicitly expose the logics of data packing so that users can easily modify it.